### ISSUE TITLE
Использование XBase для грамматики РДО

### ISSUE THREAD

**Issue by bogachev-pa** [bogachev-pa/rao-grammar-test#1]
@k-alexandrovsky
Буду коммитить в этот репозитарий наработки по грамматике.

Пока просто описал несколько упрощенную грамматику ресурсов и типов ресурсов.
https://github.com/bogachev-pa/rao-grammar-test/commit/ce8ce2a08dd780a130bedd4677f7c3294267e8ff
Зависаю на следующем: нужно как-то переводить конструктор ресурсов в джава-код. Я правильно понимаю, что он должен быть `XExpression`? Не очень представляю, как его сделать таковым.
Например, сейчас я описываю объявление ресурсов так:

``` java
ResourceDeclaration:
    'resource' name = ID '=' reference = [ResourceType | ID]
    '(' (params += XExpression (',' params += XExpression)*)? ')' ';'
;
```

Тогда не очень понятно, как получить из этого `XExpression` для конструктора.

А если делать так:

``` java
ResourceDeclaration:
    'resource' name = ID '=' constructor = RAOResourceConstructor
;

RAOResourceConstructor returns XExpression:
    {RAOResourceConstructor}
    reference = [ResourceType | ID]
    '(' (params += XExpression (',' params += XExpression)*)? ')' ';'
;
```

То непонятно, как добраться до полей `reference` и `params`, т.к. при обращении к `constructor` их не видно (видимо из-за того, что это `XExpression`, а не `EObject`).

Еще момент, стоит ли залить проект целиком или достаточно отдельных файлов?

**Comment by k-alexandrovsky** [comment]
С конструктором есть хитрый хак. Идем в грамматику `Xbase` и видим там среди всего прочего такой токен:

``` erlang
XConstructorCall returns XExpression:
    {XConstructorCall}
    'new' constructor=[types::JvmConstructor|QualifiedName] 
    (=>'<' typeArguments+=JvmArgumentTypeReference (',' typeArguments+=JvmArgumentTypeReference)* '>')?
    (=>explicitConstructorCall?='(' 
        (
            arguments+=XShortClosure
          | arguments+=XExpression (',' arguments+=XExpression)*
        )? 
    ')')?
    arguments+=XClosure?;
```

Ага, думаем мы. И без зазрения совести копируем избранное содержимое в наш рдо-конструктор, заодно опускаем `new` и вычищаем всякие странности:

``` erlang
ResourceDeclaration:
    'resource' name = ID '=' constructor = RAOResourceConstructor ';'
;

RAOResourceConstructor returns XExpression:
    {XConstructorCall}
    constructor=[types::JvmConstructor|QualifiedName]
    explicitConstructorCall?='(' 
        ( arguments+=XExpression (',' arguments+=XExpression)* )?
    ')'
;
```

После чего получаем готовенький `XExpression`, который компилится в джавовский `new Type(params)`.

И еще. В Xbase ссылки на EObject'ы (`[...|...]`) работают плохо. Там надо сразу ссылаться на джава-типы. Вот в конструкторе это `[types::JvmConstructor`, есть в этих types еще что-то вроде `JvmDeclaredType`. `types`, кстати, надо импортировать. Можно скопипастить эту строчку из исходника `Xbase.xtext`. И в правой части только `QualifiedName`, `ID` не подойдет.

**Comment by k-alexandrovsky** [comment]
Про проект и файлы не очень понял. Ты про текущий репозиторий? Заливать ли в него лишь эти два файла, и будет ли все работать?

**Comment by bogachev-pa** [comment]
> Ты про текущий репозиторий?

Да, в смысле, нужно ли в этот репозиторий заливать целый проект, который можно выкачать и запустить, или просто отдельные файлики, которые хочется посмотреть. Я думаю пока придерживаться второго, а, если вдруг понадобится, залить целиком.

**Comment by k-alexandrovsky** [comment]
Ну да, можно так. Я вообще не знаю, как объединять два проекта, на самом деле. Видимо, в старый копипастить новые файлы (ну или делать мерж). И будет огромный некрасивый коммит.

**Comment by k-alexandrovsky** [comment]
Исправил кривую копипасту в `RAOResourceConstructor`. Еще предлагаю отказаться в грамматике от префиксов `RDO`/`RAO`, мне кажется, они немного замусоривают грамматику и потом код.

**Comment by bogachev-pa** [comment]
С конструторами разобрался: https://github.com/bogachev-pa/rao-grammar-test/commit/69769ea5c7d85ea26dfee7c32da2383431f597e4.
Теперь добавил константы и собираюсь добавить функции.
1. Вопрос такой: где они должны храниться? Сейчас я просто добавляю константы туда же, куда и ресурсы, в `db`. Но тогда и обращаться к ним придется через `db`. С функциями получится то же самое. Это кажется неправильным.
   Но вроде ты говорил, что проблемы с размещением были только у ресурсов, ты как-то по другому это делал?
   Edit: Вообще это касается и последовательностей и паттернов и всего прочего.
2. Кажется, что для алгоритмических функций не нужна собственная рдо-обертка, может лучше просто разрешить писать функции на `Xtend`'е? `def` там и все дела.

**Comment by k-alexandrovsky** [comment]
1. Вообще решение проблемы может лежать в их механизме импорта. Я хотел как делать - создавать класс`имя_модели.java`, а туда кидать константы, функции, даже типы в виде вложенных классов - то есть вообще все. И потом неявно делать `import имя_модели.*` (точнее `*` там не работает, надо будет все объекты импортировать ручками). Но там вроде как сразу хрень с константами, их так просто не заимпортить. Поэтому можно сделать пакет с именем модели, туда засунуть класс `DB`. В общем, я не смог придумать правильное решение на тот момент. Можешь еще прочитать http://help.eclipse.org/juno/index.jsp?topic=%2Forg.eclipse.xtext.doc%2Fcontents%2F199-xtext-and-java.html, там есть про переменную `'it'`, может она решит наши проблемы, если удастся разобраться, как ей пользоваться. Имеется в виду передача всей `db` в конвертеры. Вот еще всякая милота про ее использование - http://blog.efftinge.de/2011/10/xtend-21-type-safe-groovy-style-builder.html.
2. Немного не понял про обертку. Имеется в виду синтаксис? Типа забить на свой и разрешить писать на xtend'е? Просто xtend это отдельный язык, у него есть своя грамматика, и тогда нам просто придется ее копипастить к себе, а это ничем не лучше. А если там еще и свои какие то улучшения для xbase написаны, то нам придется еще и xtend'овский компилятор потянуть.

**Comment by bogachev-pa** [comment]
1. Если пихать в один класс вообще всё, то никаких проблем с областями видимости не возникает, по идее, зачем вообще импорты? С другой стороны это стремно и некрасиво.
   И еще была какая-то хитрость с двухпроходным выполнением этого `accept`. Если все будет в одном классе, то так просто кросс-референсы вроде не разрешить.
2. Я так понимаю, `XBase` не предоставляет грамматики для функций, думал, их можно нахаляву получить из икстенда. Если нет, то сделаем свои, а внутри просто `XBlockExpression`.
   Cобственно, добавил функции https://github.com/bogachev-pa/rao-grammar-test/commit/f03a2c9e601e75eb8edaad8b097fef90b701524f.
   Тут еще одна загвоздка, получается, `XBase` не позволяет явно задавать тип переменной? Только `val`/`var`? Кажется как-то неудобно.
   Может имеет смысл отнаследовать и как-то переопределить грамматику `XBase`?

**Comment by k-alexandrovsky** [comment]
1. Возникает. Xbase во вложенном классе в упор не хочет видеть соседей по классу-контейнеру. Я это решал через импорт этих классов. Коряво, особенно если пользователь сам должен это писать. Если это скрыть и делать незаметно, то тогда xbase будет обманут, а джаве на такое фиолетово. Ей даже лишние импорты не вредят, покуда есть `@SuppressWarnings`. Про двухпроходность интересно звучит, теперь я, похоже, понимаю, почему даже с импортами иногда глючило. В общем, тогда нас спасет переменная db. Типы отдельно, объекты отдельно.
2. В принципе, можно задать тип и явно.

``` xtend
val/var SomeInterface object = new SomeInterfaceImpl();
```

Но можно также переопределить токены грамматики, от которой наследуешься. То есть запросто переписать правило на variable declaration (или как оно там у xbase называется), насильно заставив пользователя определять тип. Таким же примерно образом, каким родился конструктор ресурсов. Только имя токена будет таким же. Я так переопределял `ID`, у которого корни растут аж из грамматики `Terminals`, и получал юникод в идентификаторах вообще везде. Можно даже попробовать вычистить `val/var` из правила, заменив одно из них пустотой, другое традиционным `final`, хотя это может взорвать ANTLR, конечно.
